crates/cext/src/circuit.rs

```diff
@@ -874,31 +877,54 @@ pub unsafe extern "C" fn qk_circuit_get_instruction(
     let out_params = params_vec.as_mut_ptr();
     std::mem::forget(params_vec);
 
-    CInstruction {
-        name: CString::new(packed_inst.op.name()).unwrap().into_raw(),
-        num_qubits: qargs.len() as u32,
-        qubits: out_qargs,
-        num_clbits: cargs.len() as u32,
-        clbits: out_cargs,
-        num_params: params.len() as u32,
-        params: out_params,
+    // SAFETY: The pointer must point to a CInstruction size allocation
+    // per the docstring.
+    unsafe {
+        std::ptr::write(
+            instruction,
+            CInstruction {
+                name: CString::new(packed_inst.op.name()).unwrap().into_raw(),
+                num_qubits: qargs.len() as u32,
+                qubits: out_qargs,
+                num_clbits: cargs.len() as u32,
+                clbits: out_cargs,
+                num_params: params.len() as u32,
+                params: out_params,
+            },
+        );
     }
 }
 
 /// @ingroup QkCircuit
-/// Free a circuit instruction object.
+/// Clear the data in circuit instruction object.
+///
+/// This function doesn't free the allocation for the provided ``QkCircuitInstruction`` pointer, it
+/// only frees the internal allocations for the data contained in the instruction. You are
+/// responsible for allocating and freeing the actual allocation used to store a
+/// ``QkCircuitInstruction``.
 ///
-/// @param inst The instruction to free.
+/// @param inst A pointer to the instruction to free.
+///
+/// # Example
+///
+///     QkCircuitInstruction *inst = malloc(sizeof(QkCircuitInstruction));
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_gate(qc, QkGate_H, *[0], NULL);
+///     qk_circuit_get_instruction(qc, 0, inst);
+///     qk_circuit_instruction_clear(inst); // free the data
+///     free(inst); // free the pointer
+///     qk_circuit_free(qc); // free the circuit
 ///
 /// # Safety
 ///
 /// Behavior is undefined if ``inst`` is not an object returned by ``qk_circuit_get_instruction``.
```

Comment:
This should probably be something like this now?
```suggestion
/// Behavior is undefined if ``inst`` is not a valid, non-null pointer to a ``QkCircuitInstruction``.
```